GUIDELINES.md

```diff
@@ -0,0 +1,994 @@
+
+```kotlin
+// Interact with the class under test
+fun whenPassingCredentials(email: String, password: String) {
+    instance.loginOrRegister(email, password)
+}
+
+// Verify the call to the SessionStore mock
+fun thenSessionWasStarted() {
+    verify(sessionStore).beginSession(anyString().nonNullable, any<User>().nonNullable)
+}
+```
+
+#### Use the fake backend in the integration test
+
+So currently the integration test we created in the beginning still uses it's own backend mock. Let's adapt the `LoginSteps` class to use the same fake backend created previously:
+
+```kotlin
+class LoginSteps(testContext: TestContext) : BaseSteps(testContext, appModuleTestingConfiguration) {
+
+    val backend by steps<BackendFakeSteps>() // <-- added
+
+    private val viewModel by dependency<LoginViewModel>()
+//  private val backendGateway by dependency<BackendGateway>() <-- removed
+
+    override fun configure() = super.configure()
+        .requireReal<LoginViewModel>()
+//      .requireReal<AuthManager>() <-- instead of configuring it here, we just...
+        .requireSteps<AuthManagerSteps>() // <-- ...include the steps class which does the configuration
+
+    fun whenLoggingIn(email: String, password: String) {
+        viewModel.loginOrRegister(email, password)
+    }
+
+// Not needed anymore:
+
+//  fun givenExistingUser(email: String, password: String, authToken: AuthToken) {
+//      `when`(backendGateway.checkEmail(email)).thenReturn(true)
+//      `when`(backendGateway.login(email, password)).thenReturn(authToken)
+//  }
+
+
+//  fun thenEmailWasCheckedAtBackend(email: String) {
+//      verify(backendGateway).checkEmail(email)
+//  }
+
+}
+```
+
+As you can see, all the details regarding how `AuthManager` is being tested can go to the `AuthManagerSteps`, including configuration. And the integration test is still an integration test as `AuthManagerSteps` configures the `AuthManager` to be a real instance.
+
+Additionally, we add access to `backend` and again offer it as a public member so it can be used in the test. Also we use `BackendFakeUser.USER_A` in order to not only reuse test code and configuration but also test _data_! Here's the result:
+
+```kotlin
+class LoginTest : BaseJUnitTest(appModuleTestingConfiguration) {
+
+    private val sut by steps<LoginSteps>()
+
+    @Test
+    fun `Logging in checks email at backend`() = sut {
+        backend.givenExistingUser(USER_A)
+        whenLoggingIn(USER_A.email, USER_A.password)
+        backend.thenEmailWasChecked(USER_A.email)
+    }
+
+// Before:
+
+//  @Test
+//  fun `Logging in checks email at backend`() = sut {
+//      givenExistingUser(email = EXISTING_EMAIL, password = EXISTING_PASSWORD, authToken = EXISTING_AUTH_TOKEN)
+//      whenLoggingIn(email = EXISTING_EMAIL, password = EXISTING_PASSWORD)
+//      thenEmailWasCheckedAtBackend(EXISTING_EMAIL)
+//  }
+
+// Not needed anymore:
```

Comment:
kinda the same topic as above.. I would maybe focus on the changes in the test (since using the fake user is already mentioned in the text).

I don't see value in advertising the code saving in a guideline since I think everyone uses sweetest anyway/already..

but maybe it makes sense to have this comparison in a split view (if that is somehow possible) to show the difference/progress and what enden up where (than i makes sense to me to show the whole file, as a snippet I would probably focus on the test)

Reply:
hm, at mySugr yes, but this is an open-source project 😁
let's continue on the other comment...

Reply:
ah ok. wasn't aware of that :)